src/loader.ts

```diff
@@ -18,29 +18,92 @@
  * @packageDocumentation
  */
 
+import * as path from 'path';
+import * as semver from 'semver';
+import * as readPkgUp from 'read-pkg-up';
 /**
  * Import function signature type's definition.
  */
 import {HandlerFunction} from './functions';
 
+// Dynamic import function required to load user code packaged as an
+// ES module is only available on Node.js v13.2.0 and up.
+//   https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#browser_compatibility
+// Exported for testing.
+export const MIN_NODE_VERSION_ESMODULES = '13.2.0';
+
+/**
+ * Determines format of the given module (CommonJS vs ES module).
+ *
+ * Implements "algorithm" described at:
+ *   https://nodejs.org/api/packages.html#packages_type
+ *
+ * In words:
+ *   1. A module with .mjs extension is an ES module.
+ *   2. A module with .clj extension is CommonJS.
+ *   3. A module with .js extensions where the nearest package.json's
+ *      with "type": "module" is an ES module
+ *   4. Otherwise, it is CommonJS.
+ *
+ * @returns {Promise<'commonjs' | 'module'>} Module format ('commonjs' or 'module')
+ */
+async function moduleFormat(
```

Comment:
A few nits here: 

- we already have the `path` module imported so can we use the [`extname`](https://nodejs.org/api/path.html#path_path_extname_path) method
- I think it would be simpler to change the API to return a boolean and just rename it to `isESModule`
- we should avoid single line if expressions

```typescript
async function isESModule(
  modulePath: string
): Promise<boolean> {
  const ext = path.extname(modulePath);
  if (ext === '.mjs') {
    return true;
  }
  if (ext === '.cjs') {
    return false;
  }

  const pkg = await readPkgUp({
    cwd: path.dirname(modulePath),
    normalize: false,
  });

  // Default to commonjs unless package.json specifies type as 'module'.
  return pkg?.packageJson.type === 'module';
}
```

Reply:
Great points. Fixed.